File: 897-increasing-order-search-tree/897-increasing-order-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def increasingBST(self, root: TreeNode) -> TreeNode:
        
        self.dummyNode = None
        self.head = None
        
        def inorder(node):
            if node is None:
                return
            
            inorder(node.left)
            node.left = None
            
            if self.dummyNode is None:
                self.head = node
            else:
                self.dummyNode.right = node
                            
            self.dummyNode = node
            inorder(node.right)
        
        inorder(root)
        return self.head

File: 897-increasing-order-search-tree/897-increasing-order-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode head;
    TreeNode dummy;
    public TreeNode increasingBST(TreeNode root) {
        head = dummy = null;
        inorder(root);
        return head;
    }
    private void inorder(TreeNode node){
        if(node == null)
            return;
        inorder(node.left);
        node.left = null;
        
        if(dummy == null)
            head = node;
        else
            dummy.right = node;
        
        dummy = node;
        inorder(node.right);
    }
}